Day 30: “The three semantics of HTML” / Tomasz Jakut

Comandeer opened the pull request:
Adding my article, ready for the review!

Comandeer committed:
Add the "The three semantics of HTML" article.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
-Some text.
 
-Some text. Some text.
+There is always that one elephant in the room alongside HTML – _semantics_. You can ignore it for quite a while, but at the end of the day, you'd need to address that it's there, cluttering half of the space.
```
Update “you'd need to address that it's there” to “you'll need to acknowledge it's there” to better fit the metaphor.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
-Some text.
+Detailed descriptions of all HTML elements are in [chapter 4, called <cite>The elements of HTML</cite>](https://html.spec.whatwg.org/multipage/#toc-semantics).
 
+<p class="highlight"><strong>Note:</strong> That's why I'm not keen on the <q>semantic elements</q> concept. For me, they are all semantic – even <code>&lt;div&gt;</code> – as their meanings are described in the specification.</p>
```
Change “as their meanings are described in the specification.” to “since their meanings are described in the specification.”.  Makes for a smoother flow.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+<p class="highlight"><strong>Note:</strong> That's why I'm not keen on the <q>semantic elements</q> concept. For me, they are all semantic – even <code>&lt;div&gt;</code> – as their meanings are described in the specification.</p>
 
-<p class="highlight"><strong>Note:</strong> Some text.</p>
+But what if I tell you there are actually three elephants in the room?
```
Change “But what if I tell you there are actually three elephants in the room?” to “But what if I told you there are actually three elephants in the room?” to maintain proper tense.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+## First semantics: for users
+
+This is the most straightforward one of them all. Because that's the most basic role of semantics: to provide the end user with all the information necessary to correctly understand the content.
```
Change “Because that's the most basic role of semantics” to “That’s the most basic role of semantics”.  Sentences shouldn't start with "Because".

ksolomon commented on hell/adventcalendar/2025/30/index.md:
````diff
@@ -1,35 +1,212 @@
+</div>
+```
+
+It's still "semantic HTML" (because every HTML element has a meaning!), yet it's not especially useful. And yes, we can style it in any imaginable way (I bet we can even make it dance can-can with the magic of CSS animations) – but it's still just a bunch of `<div>`s [without any particular meaning](https://html.spec.whatwg.org/multipage/grouping-content.html#the-div-element):
````
Change “we can even make it dance can-can with the magic of CSS animations” to “we can even make it do the can-can with a bit of CSS magic”.  Feels more in-tone with the rest of the text.

ksolomon replied:
Change “without any particular meaning” to “without any intrinsic meaning”.  "Intrinsic" fits better semantically.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+* The image got a [descriptive `alt` attribute](https://html.spec.whatwg.org/multipage/images.html#alt), describing its content.
+* The last `<div>` was replaced with the [`<p>` element](https://html.spec.whatwg.org/multipage/grouping-content.html#the-p-element) – after all, it's a paragraph of text.
+
+If we open such a document in the browser, it looks _better_:
```
Change "If we open such a document in the browser" to "If we open the document in the browser".  More clarity on which document you're referring to.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+<img src="./a11y-div.png" alt="Both &quot;Comandeer&quot; and &quot;Proud HTML Expert&quot; are marked up as static text.">
+
+It's just a text. What's more important, the image is missing, due to its empty `alt` attribute. The `<article>` version is much better:
```
Change “What's more important, the image is missing” to “More importantly, the image is missing”.  More grammatically correct.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+<img src="./a11y-article.png" alt="The whole bio is wrapped in the article, the &quot;Comandeer&quot; is correctly identified as a heading, the image has a description of &quot;A photo of smiling Comandeer&quot; and the &quot;Proud HTML Expert&quot; is marked up as a paragraph.">
+
+Thanks to that, the screen reader can successfully transform the page content into voice and inform the user about where the articles, headings, etc, are. It would not only allow the user to fully understand the page (similarly to how the non-screen-reader user perceived the visual version of the page) but also allow the screen reader to provide better ways of navigating the page, e.g., by [jumpi ...
```
Change “inform the user about where the articles, headings, etc, are.” to “inform the user where the articles, headings, and other elements are.”.  Smoother flow.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+Thanks to that, the screen reader can successfully transform the page content into voice and inform the user about where the articles, headings, etc, are. It would not only allow the user to fully understand the page (similarly to how the non-screen-reader user perceived the visual version of the page) but also allow the screen reader to provide better ways of navigating the page, e.g., by [jumpi ...
+
+In other words: caring for semantics is caring for the user!
```
Change this to read “In other words, caring about semantics is caring for the user!” More grammatically correct, and removes the improper colon use.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+## Second semantics: for user agents
+
+Semantics can also be used to provide additional information for user agents. And I deliberately use this term in quite a wide meaning, as any machine/software that does something on behalf of its user. The most obvious example of a user agent is a browser (which is what is typically called a user agent), but I would extend that name also to search engines. With a long history of SEO, it's not a  ...
```
Change “as any machine/software that does something on behalf of its user.” to “to mean any machine or software acting on behalf of its user.” to clarify definition.

ksolomon replied:
Change "which is what is typically called a user agent" to “which is the term’s typical meaning” for smoother flow.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+In our example, the [Schema.org](https://schema.org/) vocabulary is used. Schema.org is a community effort, founded by Google, Microsoft, Yahoo, and Yandex, with a goal to create an open standard for structured data on the Web. This data is then used by search engines to provide additional information about the websites' content. [Google provides a gallery](https://developers.google.com/search/do ...
+
+The only problem with Web 3.0 is that it never happened. Granted, we have several great technologies to mark up additional metadata. But at the end of the day, they're used mainly by search engines. The ambitious vision of the Web where machines talk to each other and exchange data to provide better services to their users is still just a vision.
```
Change “mark up additional metadata” to “embed additional metadata”.  Not technically incorrrect, but "embed" fits better.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+Semantics can also be used to provide additional information for user agents. And I deliberately use this term in quite a wide meaning, as any machine/software that does something on behalf of its user. The most obvious example of a user agent is a browser (which is what is typically called a user agent), but I would extend that name also to search engines. With a long history of SEO, it's not a  ...
+
+The most ambitious take on semantics for user agents is Web 3.0. And no, I don't mean Web3, based on blockchain. Long before that, back in the 2000s, W3C imagined a new type of web, as they called it – [Web of Data and Services, aka Semantic Web](https://www.w3.org/2007/Talks/0123-sb-W3CEmergingTech/Overviewp.pdf). The idea itself was quite simple yet ingenious: the Web is full of data, but there ...
```
Change “it's not built on top of HTML as it has its own syntaxes (yes, plural).” to “it's not built on top of HTML, as it has its own syntaxes (yes, plural).”. MIssing comma.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+And finally, we can leave the vast land of user-centered semantics and enter the cozy territory of developer-centered semantics. Because an easily understandable HTML is also an easily maintainable HTML. And if we use HTML elements as intended (like creating lists with `<ul>` and `<li>` elements instead of `<divs>`), we're already halfway there!
+
+But sometimes additional info is needed, e.g., a page contains several `<article>` elements and we want to easily distinguish between them. In that case, we can go back to webdev roots and use the `class` attribute:
```
Change “use the class attribute:” to “use the class attribute.”  Don't need a colon here.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+## Third semantics: for web developers
+
+And finally, we can leave the vast land of user-centered semantics and enter the cozy territory of developer-centered semantics. Because an easily understandable HTML is also an easily maintainable HTML. And if we use HTML elements as intended (like creating lists with `<ul>` and `<li>` elements instead of `<divs>`), we're already halfway there!
```
Replace `<divs>` with `<div>`s for proper pluralization.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
````diff
@@ -1,35 +1,212 @@
+}
+```
+
+It ensures that custom elements are displayed as block elements (so like `<div>`s) as they're displayed as inline ones (so like `<span>`s) by default.
````
Change “so like `<div>`s” to “that is, like `<div>`s”.  Feels better grammatically.

ksolomon replied:
Change “as they're displayed as inline ones (so like `<span>`s)” to “because by default, they’re displayed inline (like `<span>`s)”. Better clarifies meaning.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+## Fourth semantics?
+
+Yet there is another elephant looming on the horizon… There is no doubt that we live in the AI era. LLMs are everywhere around us, and they are yet another machines that can understand HTML. But hey, they are _language models_, finally our technology has advanced enough to be able to fully grasp the beauty of Web 3.0!
```
Change "Yet there is another elephant looming on the horizon" to "Yet another elephant looms on the horizon".  Cleaner flow.

ksolomon replied:
Change “they are yet another machines” to “they are yet another kind of machine”.  Grammar fix.

ksolomon commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+Except it doesn't seem to be the case. Instead of using what we already have, new standards emerge [insert XKCD strip about standards here]. The most popular of them is the [`/llms.txt` file](https://llmstxt.org/). It's a Markdown file containing detailed instructions for LLMs on how to handle the website and where particular info is located.
+
+And I find it quite ironic that after all these years of building on the idea of the semantic Web, we finally ended up with theoretically _the_ technology to make use of all these RDFs and JSON-LDs of the Internet – and it doesn't work. Oh, well, maybe in the next 20 years, then…
```
Change "Oh, well, maybe in the next 20 years, then…" to "Oh, well, maybe in another 20 years…".  Better flow.

ksolomon replied:
First sentence feels heavy.  I recommend something like this:

"It’s quite ironic that after years of striving toward a semantic Web, we finally have _the_ technology capable of understanding it - and it still doesn’t work."

ksolomon reviewed (changes requested):
I really dug this one.  There's a few spots that grammatically or flow-wise need attention, but nothing overly "wrong", just polish and tweaks.  If you have any questions, please feel free to reach out!

matuzo commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+I'm using a [BEM-like](https://getbem.com/) naming convention here to indicate that this particular `<article>` is a bio. Each of the elements inside the `<article>` also has its own class name, indicating its purpose.
+
+<p class="highlight"><strong>Note:</strong> It can be argued that most of these classes are redundant, repeating the information already provided by RDFa. Yet I tend to treat BEM not as a naming convention but as a DSL for creating component-based applications. However, that's a story for another time.</p>
```
> as a DSL for

When you use abbreviations for the first time, please write them out.

> as a DSL (Digital Subscriber Line?) for

matuzo commented on hell/adventcalendar/2025/30/index.md:
````diff
@@ -1,35 +1,212 @@
+}
+```
+
+It ensures that custom elements are displayed as block elements (so like `<div>`s) as they're displayed as inline ones (so like `<span>`s) by default.
````
Does this really contribute to describing the third type of semantics? If so, please describe it more clearly and use another code demo, with only one element instead of seven. If not, then please remove this section.

matuzo commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,212 @@
+
+Except it doesn't seem to be the case. Instead of using what we already have, new standards emerge [insert XKCD strip about standards here]. The most popular of them is the [`/llms.txt` file](https://llmstxt.org/). It's a Markdown file containing detailed instructions for LLMs on how to handle the website and where particular info is located.
+
+And I find it quite ironic that after all these years of building on the idea of the semantic Web, we finally ended up with theoretically _the_ technology to make use of all these RDFs and JSON-LDs of the Internet – and it doesn't work. Oh, well, maybe in the next 20 years, then…
```
Have you seen this? https://help.openai.com/en/articles/12627856-publishers-and-developers-faq#h_30e9aae450

https://adrianroselli.com/2025/10/openai-aria-and-seo-making-the-web-worse.html

Do you have any thoughts? (I'm sure you do :)) Maybe it's worth mentioning.

matuzo reviewed (commented):
Thank you for this interesting article. I left you some comments.

As a reader, I'm now wondering, what does that mean for me. Do I have to look at HTML differently? Do I even have to change something about the way I develop? Why is this distinction necessary? There's a final conclusion missing.

Comandeer committed:
* Added summary
* Added comment on Atlas
* Expanded and linked DSL, SEO and LLM acronyms
* Applied suggestions from the review

Comandeer commented:
@ksolomon, @matuzo, thx for the review! I've applied your suggestions and added two additional sections – a missing conclusion and a short comment on OpenAI's browser. I've also simplified the example with custom elements.

matuzo commented on hell/adventcalendar/2025/30/index.md:
````diff
@@ -1,35 +1,200 @@
+</div>
+```
+
+It's still "semantic HTML" (because every HTML element has a meaning!), yet it's not especially useful. And yes, we can style it in any imaginable way (I bet we can even make it do the can-can with a bit of CSS magic) – but it's still just a bunch of `<div>`s [without any intrinsic meaning](https://html.spec.whatwg.org/multipage/grouping-content.html#the-div-element):
````
Can you please link can-can?

Comandeer replied:
Linked

matuzo commented on hell/adventcalendar/2025/30/index.md:
```diff
@@ -1,35 +1,200 @@
+
+<img src="./a11y-article.png" alt="The whole bio is wrapped in the article, the &quot;Comandeer&quot; is correctly identified as a heading, the image has a description of &quot;A photo of smiling Comandeer&quot; and the &quot;Proud HTML Expert&quot; is marked up as a paragraph.">
+
+Thanks to that, the screen reader can successfully transform the page content into voice and inform the user where the articles, headings, and other elements are. It would not only allow the user to fully understand the page (similarly to how the non-screen-reader user perceived the visual version of the page) but also allow the screen reader to provide better ways of navigating the page, e.g., b ...
```
```suggestion
Thanks to that, the screen reader can successfully transform the page content into voice or braille and inform the user where the articles, headings, and other elements are. It would not only allow the user to fully understand the page (similarly to how the non-screen-reader user perceived the visual version of the page) but also allow the screen reader to provide better ways of navigating the page, e.g., by [jumping to headings](https://webaim.org/projects/screenreadersurvey10/#finding).
```

matuzo commented on hell/adventcalendar/2025/30/index.md:
````diff
@@ -1,35 +1,200 @@
+```html
+<div class="card"></div>
+<!-- vs -->
+<card-></card->
````
That's not a valid tag name

Comandeer replied:
It's a [valid custom element name](https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name), according to the HTML specification:

* it starts with a lowercase letter,
* it contains only lowercase letters,
* it contains `-`.
 
Additionally, [validator does not complain](https://validator.w3.org/nu/?showsource=yes&doc=data%3Atext%2Fhtml%3Bcharset%3Dutf-8%3Bbase64%2CPCFET0NUWVBFIGh0bWw%2BPGh0bWwgbGFuZz0iZW4iPjx0aXRsZT5UZXN0IHBhZ2U8L3RpdGxlPjxjYXJkLT48L2NhcmQtPg%3D%3D).

matuzo replied:
Oh, I missed the trailing -. Is this a good practice, though?

Comandeer replied:
In case of semantics, it should be a totally transparent change – custom elements act exactly like `<div>`s. And the code could become much more readable if, instead of bunch of `<div>`s, there are more distinguishable elements.

Yet this technique is not popular and it's rarely used. I'd argue that its "rarity" influences how it's perceived – when in reality it's as good as using plain old `<div>`s.

ksolomon replied:
While I'm not a fan of how it's done here (I prefer CEs "x-" prefixed), there's nothing inherently wrong with doing it that way.

That said, I would 100% request that be changed if I saw it in a PR from one of my juniors... 😁

matuzo replied:
yeah, same. Also, it will distract some readers, but if you insist on keeping it, we'll keep it.

Comandeer replied:
Well, I’m quite used to being the only user of Custom Elements 😅 Removed that fragment.

matuzo replied:
Sorry, I only meant the `<card->` bit. Readers would be wondering if this is a typo. I would've preferred something like `<the-card>`, but, okay, it's also fine like that. :)

So, what do you think? Are we ready? Can I merge it?

Comandeer replied:
Oh, ok. So I've restored the section about CEs but renamed the element to `the-card` 👍🏻 

And I think we're ready to merge!

matuzo reviewed (commented):
Wonderful, thank you! 3 more minor changes and then we're ready to go! :)

Comandeer committed:
Update hell/adventcalendar/2025/30/index.md

Co-authored-by: Manuel Matuzovic <manuel@matuzo.at>

Comandeer committed:
Link can-can.

Comandeer committed:
Remove CE section.

Comandeer committed:
Revert "Remove CE section."

This reverts commit 2e478c8163fa9ea41920e63ad35fa0a4623056ee.

Comandeer committed:
Rename the CE.
